### PR TITLE
Upgrade to Envoy 1.13.0

### PR DESCRIPTION
**Pull request by LukeShu**
## Description

This upgrades our copy of Envoy from 1.12.2+a few patches to 1.13.0+a few patches.  The specific commit updated to is currently the tip of the `lukeshu/1.13` branch in github.com/datawire/envoy.

Envoy builds, everything generated from Envoy builds.  I've pushed the Envoy build as 
quay.io/datawire/ambassador-base:envoy-7.4f6ae4e71699d1c7ee1a6dd8be55e65bfd085b3d.opt which will be downloaded for others building from this PR so they won't have to build Envoy themselves.

It probably makes the most sense to review this commit-by-commit.  The reviewer should also verify that there has been no loss of fidelity to the Datawire patches from envoy.git tag ambassador-1.0.0 to lukeshu/1.13.

## Testing

It builds.

## Todos

I have not (yet):
 - [x] `make check-envoy`
 - [x] `make test KAT_RUN_MODE=envoy`
 - [ ] Verified whether apro.git needs adjustments

### CODE REVIEW

**Comment by LukeShu** [comment]
I got `make check-envoy` to pass:

```
out of 550 tests: 550 tests pass.
```

It's taking 211GB, but when I had it on a 250GB drive, about half of the tests refused to run (and 1 failed), because `There were tests whose specified size is too big.`.  The passing run came from a 400GB drive. 

```console
$ sudo du -h --max-depth 0 /var/lib/docker/volumes/envoy-build
206G	/var/lib/docker/volumes/envoy-build

$ df -h /var/lib/docker
Filesystem      Size  Used Avail Use% Mounted on
/dev/vdb        400G  211G  190G  53% /var/lib/docker
```

**Comment by LukeShu** [comment]
It looks like KAT passing with this is blocked by https://github.com/datawire/apro/issues/713 ?

**Comment by LukeShu** [comment]
WTAF. How is CI passing?  It shouldn't be passing.

**Comment by alphashr** [comment]
Getting 2 test failures on both `master` and also on `lukeshu/envoy-1.13` running on Kubernaut which I think is completely unrelated to Envoy upgrade:

- FAILED python/tests/test_ambassador.py::main[ConsulTest] - AssertionError: http://consultest/consultest_consul/: expected status code 503, got 200 instead with error None
- FAILED python/tests/test_scout.py::test_scout - AssertionError: test failed

**Comment by LukeShu** [comment]
> Getting 2 test failures on both `master` and also on `lukeshu/envoy-1.13` running on Kubernaut which I think is completely unrelated to Envoy upgrade:
> 
>     * FAILED python/tests/test_ambassador.py::main[ConsulTest] - AssertionError: http://consultest/consultest_consul/: expected status code 503, got 200 instead with error None
> 
>     * FAILED python/tests/test_scout.py::test_scout - AssertionError: test failed

Both of those are tests that can't be run multiple times in a row, because they don't clean up after themselves.

**Comment by LukeShu** [comment]
Assuming you ran the tests with `KAT_RUN_MODE=envoy` after merging master, looks good to me!

**Comment by alphashr** [comment]
Envoy tests report:

Three tests failed:
```
1. //test/extensions/filters/common/ratelimit:ratelimit_impl_test           FAILED in 2.0s
  /root/.cache/bazel/_bazel_root/4e47b77d977b209382cf04b3ae32963e/execroot/envoy/bazel-out/k8-dbg/testlogs/test/extensions/filters/common/ratelimit/ratelimit_impl_test/test.log


2. //test/integration:ratelimit_integration_test                            FAILED in 14.5s
  /root/.cache/bazel/_bazel_root/4e47b77d977b209382cf04b3ae32963e/execroot/envoy/bazel-out/k8-dbg/testlogs/test/integration/ratelimit_integration_test/test.log


3. //test/integration:vhds_integration_test                                 FAILED in 7.8s
  /root/.cache/bazel/_bazel_root/4e47b77d977b209382cf04b3ae32963e/execroot/envoy/bazel-out/k8-dbg/testlogs/test/integration/vhds_integration_test/test.log

```

`ratelimit_integration_test` and `ratelimit_impl_test` is expected to break because of the hard coded legacy code that we already have for ratelimit. It is expected to be fixed when we migrate to Envoy v1.14.


Ran only `vhds_integration_test` for the second time: PASSED
```
root@92ce1f18c3ce:~/envoy# time bazel test --test_env=ENVOY_IP_TEST_VERSIONS=v4only //test/integration:vhds_integration_test; 

INFO: 243 processes: 243 linux-sandbox.
INFO: Build completed successfully, 249 total actions
//test/integration:vhds_integration_test                                 PASSED in 3.4s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 249 total actions

real    1m30.908s
user    0m0.011s
sys     0m0.037s

```

**Comment by LukeShu** [comment]
Agree on the RLS tests.  The VirtualHosts Discovery Service test flake concerns me a bit.  I don't suppose you still have the log file referred to by the failing run?

File: go.mod [hunk 1]
```diff
@@ -2,20 +2,27 @@ module github.com/datawire/ambassador
 
 go 1.13
 
+// `go mod tidy` is going to try to remove `github.com/cncf/udpa`--don't let it!
+// That will break `make generate`.  Keep the github.com/cncf/udpa version
+// in-sync with the github.com/cncf/udpa/go version.
```
Comment (kflynn) [inline go.mod]: WTAF. This feels like something that'll burn us repeatedly. 🙁

File: python/ambassador/envoy/v2/v2listener.py [hunk 2]
```diff
@@ -378,6 +378,8 @@ def v2filter_authv1(auth: IRAuth, v2config: 'V2Config'):
         }
 
     if auth_info:
+        auth_info['config']['clear_route_cache'] = True
```
Comment (kflynn) [inline python/ambassador/envoy/v2/v2listener.py:381]: Hmmmm. Why is this needed now?

File: cxx/envoy.mk [hunk 3]
```diff
@@ -14,7 +14,7 @@ IS_PRIVATE ?= $(findstring private,$(_git_remote_urls))
   ENVOY_COMPILATION_MODE ?= opt
   # Increment BASE_ENVOY_RELVER on changes to `docker/base-envoy/Dockerfile`, or Envoy recipes.
   # You may reset BASE_ENVOY_RELVER when adjusting ENVOY_COMMIT.
-  BASE_ENVOY_RELVER ?= 2
+  BASE_ENVOY_RELVER ?= 1
```
Comment (kflynn) [inline cxx/envoy.mk:17]: This seems odd?

File: go.mod [hunk 4]
```diff
@@ -2,9 +2,49 @@ module github.com/datawire/ambassador
 
 go 1.13
 
-// `go mod tidy` is going to try to remove `github.com/cncf/udpa`--don't let it!
-// That will break `make generate`.  Keep the github.com/cncf/udpa version
-// in-sync with the github.com/cncf/udpa/go version.
+// If you're editing this file, there's a few things you should know:
```
Comment (kflynn) [inline go.mod]: Shouldn't we keep that comment about `go mod tidy`? Or is it no longer relevant?

**Comment by alphashr** [comment]
> Agree on the RLS tests. The VirtualHosts Discovery Service test flake concerns me a bit. I don't suppose you still have the log file referred to by the failing run?

Because it passed on the 2nd tun, I didn't grab the log failure file offline.
This test flakiness is seen [[1](https://envoyproxy.slack.com/archives/CEFDKQ3RQ/p1558048872065300)], [[2](https://envoyproxy.slack.com/archives/CEFDKQ3RQ/p1558117767001800)] and [[3](https://github.com/envoyproxy/envoy/issues/9784)]: failure on thread sanitizer run

File: cxx/envoy.mk (hunk 3 above)
Comment (alphashr) [reply cxx/envoy.mk:17]: `478433e447e1bef3052c71b046161eb6bdab6abd` commit changed ENVOY_COMMIT from `a54d6eb82e8e83d65e52307e89557a07c0b6d1b0`  to `33ff697fc665b25463c7d64916ee6736782e2e76`. In the subsequent commit `478433e447e1bef3052c71b046161eb6bdab6abd` that merged master to this branch the `ENVOY_COMMIT` kept as previous commit but it just changed the `BASE_ENVOY_RELVER ?= 2` to `BASE_ENVOY_RELVER ?= 1`.   The Quay ambassador-base image [tagged](https://quay.io/repository/datawire/ambassador-base?tab=tags) as: `envoy-1.33ff697fc665b25463c7d64916ee6736782e2e76.opt`

File: python/ambassador/envoy/v2/v2listener.py (hunk 2 above)
Comment (alphashr) [reply python/ambassador/envoy/v2/v2listener.py:381]: clear_route_cache:
// Clears route cache in order to allow the external authorization service to correctly affect routing decisions. Filter clears all cached routes when: 
1. The field is set to *true*.
2. ...

More [here](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto#L63).

File: go.mod (hunk 4 above)
Comment (alphashr) [reply go.mod]: @LukeShu any opinion on this?

File: cxx/envoy.mk (hunk 3 above)
Comment (LukeShu) [reply cxx/envoy.mk:17]: "You may reset BASE_ENVOY_RELVER when adjusting ENVOY_COMMIT."

File: go.mod (hunk 4 above)
Comment (LukeShu) [reply go.mod]: Good question

Comment (LukeShu) [reply go.mod]: Yes, the comment is still applicable.  It seems that it was dropped in the 8918d359dc51d5b7ead5fbae93620f51d7ba1b97 merge.

**Comment by LukeShu** [comment]
And also between the fixed RLS patch and the bump to `google.golang.org/grpc` from master, this is due for another `make generate`.

File: go.mod (hunk 4 above)
Comment (LukeShu) [reply go.mod]: No, I tried that; it didn't work right :shrug:

File: python/ambassador/envoy/v2/v2listener.py (hunk 2 above)
Comment (LukeShu) [reply python/ambassador/envoy/v2/v2listener.py:381]: Because we used to apply a patch to Envoythat always cleared the route cache, from before the 'clear_route_cache' flag landed in Envoy.

Comment (kflynn) [reply python/ambassador/envoy/v2/v2listener.py:381]: @LukeShu's answer is what I needed. 🙂  Thanks!

File: cxx/envoy.mk (hunk 3 above)
Comment (kflynn) [reply cxx/envoy.mk:17]: OK, thanks!

**Comment by kflynn** [comment]
I just revoked my approval not because of a sudden issue with the PR, but because I need to resolve a question around its release vehicle. Hold up. :)
